File: Utilities/Space.m
function [AllOp,OpSpace,ParaSpace,ParaLocalSpace,BehavSpace] = Space(Problem,Setting)
% Define the design space.

% operater space
switch Problem(1).type{1}
    case 'continuous'
        Choose  = {'choose_traverse';'choose_tournament';'choose_roulette_wheel';'choose_cluster'}; 
        Search  = {'search_pso';'search_de_current';'search_de_current_best';'search_de_random';
            'cross_arithmetic';'cross_sim_binary';'cross_point_one';'cross_point_two';
            'cross_point_uniform';'search_mu_gaussian';'search_mu_cauchy';'search_mu_polynomial';
            'search_mu_uniform';'search_eda';'search_cma';'reinit_continuous'};
        Update  = {'update_greedy';'update_round_robin';'update_pairwise';'update_always';'update_simulated_annealing'};
       
    case 'discrete'
        Choose  = {'choose_traverse';'choose_tournament';'choose_roulette_wheel'};
        Search  = {'cross_point_one';'cross_point_two';'cross_point_uniform';'search_reset_one';
            'search_reset_rand';'reinit_discrete'};
        Update  = {'update_greedy';'update_round_robin';'update_pairwise';'update_always';'update_simulated_annealing'};

    case 'permutation'
        Choose  = {'choose_traverse';'choose_tournament';'choose_roulette_wheel'};
        Search  = {'cross_order_two';'cross_order_n';'search_swap';'search_swap_multi';
            'search_scramble';'search_insert';'reinit_permutation'};
        Update  = {'update_greedy';'update_round_robin';'update_pairwise';'update_always';'update_simulated_annealing'};
end
OpSpace = [1,length(Choose);
    length(Choose)+1,length(Choose)+length(Search);
    length(Choose)+length(Search)+1,length(Choose)+length(Search)+length(Update)]; % each row reports the search space of one kind of operaters, e.g., the search space of Choose is intergers from 1 to 4.
AllOp = [Choose;Search;Update];

% parameter and behavior space
ParaSpace      = cell(length(AllOp),1); % parameter space
ParaLocalSpace = cell(length(AllOp),1); % parameter space for performing local search
BehavSpace     = cell(length(AllOp),1); % search behaviors
for i = 1:length(AllOp)
    [thisParaSpace,~]  = feval(str2func(AllOp{i}),Problem,'parameter');
    [thisBehavSpace,~] = feval(str2func(AllOp{i}),'behavior');
    ParaSpace{i} = thisParaSpace;
    BehavSpace{i} = thisBehavSpace;
    if ~isempty(thisBehavSpace{1,1}) && ~isempty(thisParaSpace) % if the operator can perform local search and has parameter(s)
        if isempty(thisBehavSpace{2,1}) % if only performs local search
            thisParaLocalSpace = thisParaSpace;
        else
            thisParaLocalSpace = zeros(size(thisParaSpace));
            for j = 1:size(thisParaLocalSpace,1) % for each parameter of operator i
                lower = thisParaSpace(j,1);
                upper = thisParaSpace(j,2);
                trend = thisBehavSpace{1,j+1}; % parameter setting for local search
                if strcmp(trend,'small')
                    thisParaLocalSpace(j,:) = [lower,lower+(upper-lower)*Setting.LSRange];
                elseif strcmp(trend,'large')
                    thisParaLocalSpace(j,:) = [upper-upper*Setting.LSRange,upper];
                end
            end
        end
        ParaLocalSpace{i} = thisParaLocalSpace;
    end
end

File: src/autooptlib/utils/space.py
from __future__ import annotations

from types import SimpleNamespace
from typing import Any

import numpy as np

from ..components import get_component
from .design._helpers import get_problem_type


def _to_namespace(setting: Any) -> SimpleNamespace:
    if isinstance(setting, SimpleNamespace):
        return setting
    if isinstance(setting, dict):
        return SimpleNamespace(**setting)
    # Fallback: copy attributes
    data = {k: getattr(setting, k) for k in dir(setting) if not k.startswith("__") and not callable(getattr(setting, k))}
    return SimpleNamespace(**data)


def _to_behavior_matrix(behavior: Any) -> list[list[Any]] | None:
    """Convert component behavior metadata to a list-of-lists structure."""
    if behavior is None:
        return None
    if isinstance(behavior, np.ndarray):
        behavior = behavior.tolist()
    elif isinstance(behavior, tuple):
        behavior = list(behavior)
    elif not isinstance(behavior, list):
        return None

    matrix: list[list[Any]] = []
    for row in behavior:
        if isinstance(row, np.ndarray):
            matrix.append(row.tolist())
        elif isinstance(row, (list, tuple)):
            matrix.append(list(row))
        elif row is None:
            matrix.append([])
        else:
            matrix.append([row])
    return matrix


def _compute_local_bounds(behavior: list[list[Any]] | None, bounds: np.ndarray | None, ls_range: float) -> np.ndarray | None:
    """Compute the parameter subspace used for local search, mirroring MATLAB Space.m."""
    if behavior is None or bounds is None or bounds.size == 0:
        return None

    if not behavior or not behavior[0]:
        return None
    first_col = behavior[0][0]
    if first_col in (None, "", 0):
        return None

    has_global = len(behavior) > 1 and behavior[1] and behavior[1][0] not in (None, "", 0)
    if not has_global:
        return bounds.copy()

    ls_range = float(ls_range)
    ls_range = max(0.0, min(1.0, ls_range))
    local_bounds = np.array(bounds, copy=True, dtype=float)
    span = bounds[:, 1] - bounds[:, 0]
    trends = behavior[0][1:] if len(behavior[0]) > 1 else []

    for idx in range(bounds.shape[0]):
        lower = bounds[idx, 0]
        upper = bounds[idx, 1]
        trend = trends[idx] if idx < len(trends) else None
        if trend == "small":
            local_bounds[idx, 0] = lower
            local_bounds[idx, 1] = lower + span[idx] * ls_range
        elif trend == "large":
            local_bounds[idx, 0] = upper - span[idx] * ls_range
            local_bounds[idx, 1] = upper
        else:
            local_bounds[idx, 0] = lower
            local_bounds[idx, 1] = upper

    return local_bounds


def space(problem: Any, setting: Any) -> SimpleNamespace:
    """Define operator/parameter space for the design problem (continuous only)."""
    set_obj = _to_namespace(setting)
    ls_range = float(getattr(set_obj, "LSRange", 0.25))
    set_obj.LSRange = ls_range
    ptype = get_problem_type(problem)

    if ptype == "continuous":
        choose = ["choose_traverse", "choose_tournament", "choose_roulette_wheel", "choose_nich"]
        search = [
            "search_de_current",
            "search_de_current_best",
            "search_de_random",
            "search_cma",
            "search_pso",
            "cross_point_one",
            "cross_point_two",
            "cross_point_uniform",
            "cross_point_n",
            "search_mu_gaussian",
            "search_mu_cauchy",
            "search_mu_polynomial",
            "search_mu_uniform",
            "reinit_continuous",
        ]
        update = ["update_greedy", "update_round_robin", "update_pairwise", "update_always", "update_simulated_annealing"]
    elif ptype == "discrete":
        choose = ["choose_traverse", "choose_tournament", "choose_roulette_wheel", "choose_nich"]
        search = [
            "cross_point_one",
            "cross_point_two",
            "cross_point_uniform",
            "cross_point_n",
            "search_reset_one",
            "search_reset_rand",
            "search_reset_creep",
            "reinit_discrete",
        ]
        update = ["update_greedy", "update_round_robin", "update_pairwise", "update_always", "update_simulated_annealing"]
    elif ptype == "permutation":
        choose = ["choose_traverse", "choose_tournament", "choose_roulette_wheel", "choose_nich"]
        search = [
            "cross_order_two",
            "cross_order_n",
            "search_swap",
            "search_swap_multi",
            "search_scramble",
            "search_insert",
            "reinit_permutation",
        ]
        update = ["update_greedy", "update_round_robin", "update_pairwise", "update_always", "update_simulated_annealing"]
    else:
        raise NotImplementedError("space() currently supports continuous, discrete, and permutation problems")

    all_op = choose + search + update
    op_space = np.array(
        [
            [1, len(choose)],
            [len(choose) + 1, len(choose) + len(search)],
            [len(choose) + len(search) + 1, len(all_op)],
        ],
        dtype=int,
    )

    para_space = []
    behav_space = []
    para_local_space = []

    for name in all_op:
        arr = None
        try:
            comp = get_component(name)
        except KeyError:
            para_space.append(None)
            behav_space.append(None)
            para_local_space.append(None)
            continue

        try:
            params, _ = comp(problem, "parameter")
        except TypeError:
            params = None
        try:
            behavior, _ = comp("behavior")
        except TypeError:
            behavior = None

        if params is None:
            arr = None
        else:
            arr = np.asarray(params, dtype=float)
            if arr.ndim == 1:
                if arr.size % 2 != 0:
                    raise ValueError(f"Parameter bounds for {name} must contain pairs")
                arr = arr.reshape(-1, 2)
            elif arr.ndim > 1 and arr.shape[1] != 2:
                total = arr.size
                if total % 2 != 0:
                    raise ValueError(f"Parameter bounds for {name} must contain pairs")
                arr = arr.reshape(-1, 2)

        behav_matrix = _to_behavior_matrix(behavior)
        para_space.append(arr)
        behav_space.append(behav_matrix if behav_matrix is not None else behavior)
        para_local_space.append(_compute_local_bounds(behav_matrix, arr, ls_range))

    set_obj.AllOp = all_op
    set_obj.OpSpace = op_space
    set_obj.ParaSpace = para_space
    set_obj.BehavSpace = behav_space
    set_obj.ParaLocalSpace = para_local_space

    # Provide defaults for downstream code
    set_obj.TunePara = getattr(set_obj, "TunePara", False)
    set_obj.alg_p = getattr(set_obj, "alg_p", 1)
    set_obj.alg_q = getattr(set_obj, "alg_q", 3)
    set_obj.alg_n = getattr(set_obj, "alg_n", 1)
    return set_obj
